Add runbooks (#92)
mulesoft/docs-service-mesh: modules/ROOT/pages/troubleshoot-service-binding-issues.adoc

* <<dns-name-error,Binding name not in DNS subdomain standard format>>
* <<references-non-existent-instance-error,References nonexistent instance error>>
* <<error-instance-not-ready,Error instance not ready>>

[[dns-name-error]]

> == Binding Name not in DNS Subdomain Standard Format

Reviewer: See suggestions above and make the presentation (or not) of the links and titles consistent.

Reviewer 2: Done! Thanks!

Delete the erroneous binding command and re-create it with the correct input.

[[error-instance-not-ready]]

> == Error instance not ready

Reviewer: ```suggestion
== Error Instance Not Ready
```

Reviewer 2: Done! Thanks!

Before you create a binding, the API provisioning task must be completed and the referenced API or adapter must be ready. Otherwise, you receive the 'ErrorInstanceNotReady' error.

=== Diagnose

To diagnose this issue, check the status of the referenced instance by using either of the following commands, depending on whether you are using an API or an adapter binding:

* `$ asmctl api list`
* `$ asmctl adapter list`

=== Solution

To resolve this issue:

. If the adapter or API provisioning has not yet completed (`Provisioning` status), wait until the operation completes. 
. If the status is `Failed`, review the corresponding troubleshooting section:

* xref:troubleshoot-adapter-provisioning-issues.adoc[Adapter provisioning issues]
* xref:troubleshoot-api-provisioning-issues.adoc[API provisioning issues].
+
. To delete the API or adapter, delete the binding first. 

> . Wait for the API or the adapter to be ready (Not Deleting).

Reviewer: ```suggestion
. Wait for the API or the adapter to be ready (`Not Deleting`).
```

Reviewer 2: Done! Thanks!

Delete the erroneous binding command and re-create it with the correct input.

[[error-instance-not-ready]]
== Error instance not ready

When you create a binding, the referenced API or adapter must be ready. Otherwise, you receive the 'ErrorInstanceNotReady' error.

> == Causes

Reviewer 3: I think you missed one level

```suggestion
=== Causes
```

Author: ha, copied and pasted from another place, you are right!

This error can occur if:
 
* You are trying to create an API binding and the API is not ready. This can be because the API provisioning failed or has not finished yet. 
* You are trying to create an adapter binding and the adapter is not ready. This can be because the adapter provisioning failed or has not finished yet.

=== Diagnose

Diagnose this issue by running one of the following commands:

* `$ asmctl api binding list`
* `$ asmctl adapter binding list`
+
The status of the binding displays as `ErrorInstanceNotReady`. Check the status of the referenced instance by running one of the following commands (depending if you are using api or adapter binding)

> * `$ asmctl api list`

Reviewer 3: Wrong command, maybe:

```suggestion
* `$ asmctl api logs --name=<replace_with_api_name> --namespace=<replace_with_namespace>`
```

Author: In this case, the only thing I want them to notice is that the API has failed. On the next section, they have the links for API troubleshooting. I prefer not to explain the reasons for which the API failed on this runbook.

> * `$ asmctl adapter list`

Reviewer 3: Wrong command, maybe:

```suggestion
* `$ asmctl adapter logs --name=<replace_with_adapter_name> --namespace=<replace_with_namespace>`
```

Author: Same as above. In this case, the only thing I want them to notice is that the API has failed. On the next section, they have the links for API troubleshooting. I prefer not to explain the reasons for which the API failed on this runbook.

> 

Reviewer 3: Maybe you should need to add here the same comment about using `kubectl describe serviceinstances`:

```
In case there is no information above, you will have to run `kubectl describe serviceinstances <replace_with_api_name> -n <replace_with_namespace>`. In the `status` section of the resource, you will find the error message explaining the issue.
```

Author: Same as above.

=== Solution

> If the adapter or API provisioning has not finished yet(status is `provisioning`, then wait until the operation finishes. If the status is failed, go to the corresponding troublshooting section:

Reviewer 3: ```suggestion
If the adapter or API provisioning has not finished yet (status is `Provisioning`), then wait until the operation finishes. If the status is failed, go to the corresponding troubleshooting section:
```

> If the adapter or API provisioning has not finished yet (status is `Provisioning`, then wait until the operation finishes. If the status is failed, go to the corresponding troubleshooting section:

Reviewer 3: ```suggestion
If the adapter or API provisioning has not finished yet (status is `Provisioning`), then wait until the operation finishes. If the status is `Failed`, go to the corresponding troubleshooting section:
```

Reviewer 2: Done! Thanks!

Delete the erroneous binding command and re-create it with the correct input.

[[error-instance-not-ready]]
== Error instance not ready

> Before you create a binding, the referenced API or adapter must be ready. Otherwise, you receive the 'ErrorInstanceNotReady' error.

Reviewer: ditto re. title wording and capitalization: Either make all typical titles, with proper title-style caps., or make clear that you are using the actual error syntax, which here looks like all one term: the 'ErrorInstanceNotReady' error.

Reviewer: ```suggestion
== Error instance not ready

Before you create a binding, the referenced API or adapter must be ready. 
```

Reviewer: But what does being ready entail? Why would these not be ready?

Reviewer 2: Explained in the Causes section.

> === Causes

Reviewer: This is all repetitive of the error description above.

Reviewer 2: Good point! Removed the Causes section.

This error can occur if you are attempting to create:
 

> * An API binding and the API is not ready. 

Reviewer: ```suggestion
* An API binding and the API is not ready
```

Reviewer 2: Removed the sentence. Done! Thanks!

+
This can occur because the API provisioning failed or has not yet completed. 

> * An adapter binding and the adapter is not ready. 

Reviewer: ```suggestion
* An adapter binding and the adapter is not ready
```

Reviewer 2: Removed the sentence. Done! Thanks!

+
This can occur because the adapter provisioning failed or has not yet completed.

> 

Reviewer: GOING FORWARD...:

At some point, we should reevaluate the use of this troubleshooting template if it continues to result in so much repetition.

This entire set of subtopics could easily be replaced by troubleshooting text such as, in this case:

ErrorInstanceNotReady Error

"This error occurs when you attempt to create a binding and the associated API or adapter provisioning either failed or has not yet finished.

"If the adapter or API provisioning status is still `Provisioning`, wait until the operation completes to create the binding. 

"If the status is `Failed`, review the corresponding troubleshooting section:

* xref:troubleshoot-adapter-provisioning-issues.adoc[Adapter Provisioning Issues]
* xref:troubleshoot-api-provisioning-issues.adoc[API Provisioning Issues].

=== Diagnose

To diagnose this issue, run one of the following commands:

* `$ asmctl api binding list`
* `$ asmctl adapter binding list`
+

> The status of the binding displays as `ErrorInstanceNotReady`. 

Reviewer: Why do we need this text? The audience has already received the error message, which is the same as the binding status? Seems repetitive.

Reviewer 2: Done! Thanks!

> Check the status of the referenced instance using any of the following commands, depending on whether you are using an API or adapter binding:

Reviewer: ```suggestion
Check the status of the referenced instance by using either of the following commands, depending on whether you are using an API or an adapter binding:
```

Reviewer 2: Done! Thanks!

* `$ asmctl api list`

> * `$ asmctl adapter list`

Reviewer: What do these commands show the user that is different from the previous two?
Again, the user already knows that the instance is not ready, so why go through all this to doublecheck?

Reviewer 2: Leaving it in! Done! Thanks!

=== Solution

> If the adapter or API provisioning has not yet completed (`Provisioning` status), wait until the operation completes. If the status is `Failed`, review the corresponding troubleshooting section:

Reviewer: See above suggestion, in the GOING FORWARD comment box.

Reviewer 2: I think there are errors where there is a need for us to explain why an error occurred (Cause), how to diagnose what happened (Diagnose) and resolve it (solution). However, there are some errors that might be pretty straight forward. We'd need to make sure that we don't add repetitive information just to be able to fit the template.

* xref:troubleshoot-adapter-provisioning-issues.adoc[Adapter provisioning issues]
* xref:troubleshoot-api-provisioning-issues.adoc[API provisioning issues].

> To delete the API or adapter, delete the binding first. After the API or the adapter is ready, you can recreate the binding.

Reviewer: Why is this text here? Under what circumstances must the user actually delete the API or adapter rather than waiting or following the processes outlined in the cross-ref'd topics?

Reviewer: Also, this text does not make sense. Is the goal to delete the API or adapter, or to replace them?
Seems like the steps should be...:
1. Delete binding.
2. Wait for API or adapter to be ready. (NOT DELETING)
3. Re-create binding.

Reviewer 2: Done! Thanks!
